sdk/cosmos/azure_data_cosmos/src/routing/async_cache.rs

```diff
@@ -0,0 +1,217 @@
+use async_lock::RwLock;
+use std::collections::HashMap;
+use std::hash::Hash;
+use std::sync::Arc;
+use std::time::{Duration, Instant};
+
+/// Cache entry with TTL tracking
+#[derive(Clone, Debug)]
+struct CacheEntry<V> {
+    value: V,
+    expires_at: Instant,
+}
+
+impl<V> CacheEntry<V> {
+    fn new(value: V, ttl: Duration) -> Self {
+        Self {
+            value,
+            expires_at: Instant::now() + ttl,
+        }
+    }
+
+    fn is_expired(&self) -> bool {
+        Instant::now() >= self.expires_at
+    }
+}
+
+/// A generic async cache with TTL support and expiration callbacks
+#[derive(Clone)]
+pub struct AsyncCache<K, V>
+where
+    K: Eq + Hash + Clone,
+    V: Clone,
+{
+    store: Arc<RwLock<HashMap<K, CacheEntry<V>>>>,
+    ttl: Duration,
+}
+
+impl<K, V> AsyncCache<K, V>
+where
+    K: Eq + Hash + Clone + Send + Sync + 'static,
+    V: Clone + Send + Sync + 'static,
+{
+    /// Creates a new `AsyncCache` with the specified TTL and expiration callback
+    pub fn new(ttl: Duration) -> Self {
+        Self {
+            store: Arc::new(RwLock::new(HashMap::new())),
+            ttl,
+        }
+    }
+
+    /// Gets a value from the cache, or computes it using the provided async function if not present or expired
+    ///
+    /// When the entry is expired, the expiration callback is invoked before computing the new value.
+    /// The cache is automatically updated with the computed value.
+    pub async fn get<F, Fut, E>(&self, key: K, compute: F) -> Result<V, E>
+    where
+        F: FnOnce() -> Fut,
+        Fut: std::future::Future<Output = Result<V, E>>,
+    {
+        // Fast path: check if value exists and is not expired (read lock)
+        {
+            let store = self.store.read().await;
+            if let Some(entry) = store.get(&key) {
+                if !entry.is_expired() {
+                    return Ok(entry.value.clone());
+                }
+            }
+        }
+
+        // Slow path: value missing or expired, need to compute (write lock)
+        let mut store = self.store.write().await;
+
+        // Double-check after acquiring write lock (another task might have updated it)
+        if let Some(entry) = store.get(&key) {
+            if entry.is_expired() {
+                // Remove the entry from the cache.
+                store.remove(&key);
+            } else {
+                // Another task updated it while we waited for the lock
+                return Ok(entry.value.clone());
+            }
+        }
+
+        // Release lock while computing the new value
+        drop(store);
```

Comment:
This will cause concurrency issues. I think we need to keep the lock held while computing the new value. Since we're using an async-friendly lock that should be OK, all the other callers to `get` will block waiting for the same value to be computed.

Reply:
Looks like, removing this in the `get` method, would cause a deadlock. After acquiring the write lock on line `77` by calling `self.store.write().await`, if the code doesn't drop it before calling `compute().await` on line `100`, then on line `96`, it will try to acquire the write lock again with `self.store.write().await.insert(key, entry)`, which will cause a deadlock because the lock is still held.

Reply:
Yes, seems like just doing `store.insert` without releasing/ dropping the lock should work as expected. Refactored the code to reflect the change.